### DISCUSSION TITLE
Feature proposal: Ability to create more specific groups inside of `require-dev` and `dev

### DISCUSSION THREAD

**Discussion by jacobbednarz** [composer/composer#9923] [General]
Coming from a package manager like Ruby's [bundler](https://bundler.io/), one of the features I'm often missing is the ability to be more specific about my `require`s and `require-dev`s instead of the current all or nothing approach. I think it would be benefitical to introduce a similar concept to include or exclude "groups" of dependencies via CLI `install` flags.

Let's consider an example scenario where you have dev, test, stage and production. In all environments you have pretty well the same amount of runtime dependencies being loaded in `require` however in dev and stage, you now also need some internal tooling to help diagnose some issues. This functionality is potentially dangerous to run in production so you don't want to include it all by default. Similarly, in the test environment, you have runtime dependencies but you don't need or want to include them in all environments as they are solely for running tests. In the current landscape, `require-dev` is often (ab)used to handle the last scenario and while it works, it doens't lend itself to much flexibility.

Right now, it's pretty messy to handle this and the current Composer file design doesn't really facilitate it (which is expected as it hasn't been a need) but in the future, I think we could improve these types of scenarios by including an option such as `--with` or `--without` for the install commands that would faciliate being more flexible about what to install. 

I haven't though too much about the design of the `composer.json` file however to make it backwards compatible, the object keys could remain prefixed with `require_` and the `--with`/`--without` could use them. (Eventually thought we'd rip off the bandaid and rename the keys to be more informative with the `require` prefix and potentially nested) 

For example:

```json
{
    "require": {
        "example/example-dep-for-all": "*"
    },
    "require-dev": {
        "example/dev-1": "*",
        "example/dev-2": "*",
        "example/dev-3": "*"
    },
    "require-stage": {
        "example/stage-1": "*",
        "example/stage-2": "*",
        "example/stage-3": "*"
    },
}
```

And an invocation for only the staging setup.

```
composer install --without "dev"
```

Alternatively, we could also continue the path of the current set of flags with the `--no` prefix (such as `--no-dev). 

Thoughts?

**Comment by jacobbednarz** [comment]
Linking #9636 as it looks to be a similar-ish issue that would also benefit from this improvement however in a more general fashion. I think defining the schema based on the proposed fields is eventually going to run into the same issues once a new category of groups is generated.

**Reply by Seldaek** [reply]
This has been discussed in the past and IMO there is close to zero chance we will add more of `require-dev`-style groups. You may be able to find an old closed issue if you look enough :)
